src/components/graph/ExerciseTreeGraph.tsx

```diff
@@ -0,0 +1,496 @@
+      .style('stroke-width', 1)
+
+    legend
+      .append('circle')
+      .attr('r', circleRadius)
+      .attr('cx', 20)
+      .attr('cy', 20)
+      .style('fill', 'lightsteelblue')
+      .style('stroke', 'black')
+      .style('stroke-width', 1)
+
+    legend
+      .append('rect')
+      .attr('width', squareLength)
+      .attr('height', squareLength)
+      .attr('rx', squareCornerRadius)
+      .attr('ry', squareCornerRadius)
+      .attr('x', 20 - squareLength / 2)
+      .attr('y', 50 - squareLength / 2)
+      .style('fill', 'lightpink')
+      .style('stroke', 'black')
+      .style('stroke-width', 1)
+
+    legend
+      .append('text')
+      .style('font-size', 14)
+      .style('fill', '#000')
+      .text('implemented exercises')
+      .attr('x', 35)
+      .attr('y', 25)
+
+    legend
+      .append('text')
+      .style('font-size', 14)
+      .style('fill', '#000')
+      .text('concepts missing an exercise')
+      .attr('x', 35)
+      .attr('y', 55)
+
+    legend
+      .append('rect')
+      .attr('x', 1)
+      .attr('y', 1)
+      .attr('height', 70)
+      .attr('width', 225)
+      .attr('class', 'overlay')
+      .style('fill', 'rgba(0,0,0,0)')
+
+    showLegend.on('mouseover', () => {
+      d3.select('g.legend').attr('visibility', 'visible')
+    })
+
+    d3.select('g.legend > rect.overlay').on('mouseout', () => {
+      d3.select('g.legend').attr('visibility', 'hidden')
+    })
+  }
+
+  public componentWillUnmount(): void {
+    const circles = Array.from(
+      document.querySelectorAll<SVGCircleElement>('svg#concept-map circle')
+    )
+    for (const circle of circles) {
+      circle.removeEventListener('mouseover', handleCircleMouseover)
+      circle.removeEventListener('mouseout', handleCircleMouseout)
+    }
+
+    d3.select('g.show-legend').on('mouseover', null)
+    d3.select('g.legend > rect.overlay').on('mouseout', null)
+  }
+
+  public render(): JSX.Element {
+    if (!this.props.config) {
+      return <CheckOrCross value={false} />
+    }
+
+    return (
+      <div className="text-center">
+        <div id="graph-container" ref={this.graphRef} />
+      </div>
+    )
+  }
+}
+
+/**
+ * displayNoExercises
+ */
+function displayNoExercises(container: HTMLDivElement | null): void {
+  d3.select(container).append('p').text('No exercises to display as tree.')
+}
+
+/**
+ * Use function to retrieve map value so as to not have to handle potential undefined in graph block
+ */
+function get_position_from(
+  map: Map<slug | string, position>,
+  key: slug | string
+): position {
+  const value = map.get(key)
+  if (value) return value
+  throw new Error("key doesn't exist")
+}
+
+/**
+ * handleCircleMouseover
+ * This event handler highlights the current circle, paths leading from (source) and into (target)
+ * and dims all of the rest.
+ */
+function handleCircleMouseover(event: MouseEvent): void {
+  // highlight circle on mouseover
+  const targetCircle = event.target as SVGCircleElement
```

Comment:
I don't recall, but I think that `React.MouseEvent` actually is generic (defaults to `Element`), so instead you can do:

```typescript
function handleCircleMouseover(event: MouseEvent<SVGCircleElement>): void {
```

Otherwise, this is fine to keep.

Reply:
`Type 'MouseEvent' is not generic.ts(2315)` error so will keep as is